rclcpp/include/rclcpp/loaned_message.hpp

```diff
@@ -183,14 +183,30 @@ class LoanedMessage
   /**
    * A call to `release()` will unmanage the memory for the ROS message.
    * That means that the destructor of this class will not free the memory on scope exit.
+   * If the message is loaned from the middleware but not be published, the user needs to call
+   * `rcl_return_loaned_message_from_publisher` manually.
+   * If the memory is from the local allocator, the memory is freed when the unique pointer
+   * goes out instead, and the unique pointer will keep the publisher alive until it is destroyed.
    *
-   * \return Raw pointer to the message instance.
+   * \return std::unique_ptr to the message instance.
    */
-  MessageT * release()
+  std::unique_ptr<MessageT, std::function<void(MessageT *)>>
+  release()
   {
     auto msg = message_;
     message_ = nullptr;
-    return msg;
+
+    if (pub_.can_loan_messages()) {
+      return std::unique_ptr<MessageT, std::function<void(MessageT *)>>(msg, [](MessageT *) {});
+    }
+
+    return std::unique_ptr<MessageT, std::function<void(MessageT *)>>(
+      msg,
+      [ = ](MessageT * msg_ptr) {
+        // call destructor before deallocating
+        msg_ptr->~MessageT();
+        message_allocator_.deallocate(msg_ptr, 1);
+      });
```

Comment:
I would be more explicit about what is being captured, I'm not sure if you meant to use:

```suggestion
      [this](MessageT * msg_ptr) {
        // call destructor before deallocating
        msg_ptr->~MessageT();
        message_allocator_.deallocate(msg_ptr, 1);
      });
```

or

```suggestion
      [allocator=message_allocator_](MessageT * msg_ptr) {
        // call destructor before deallocating
        msg_ptr->~MessageT();
        allocator.deallocate(msg_ptr, 1);
      });
```

which by the way the difference doesn't matter, but I prefer being more explicit.

Reply:
Updated. :clap: